### PR TITLE
[Refactor] Extract DisaggregatedResult from MetricFrame

### PR DESCRIPTION
**Pull request by riedgar-ms**
## Description

The `MetricFrame` class has long had a smaller class within it attempting to escape. This long desired goal is achieved herein.

We introduce a `DisaggregatedResult` class which provides the core functionality of `MetricFrame` but in a more constrained form. Key points:

- The functions must be supplied in the form of a dictionary of `AnnotatedMetricFunction`s
- The data must all be in a `DataFrame` which contains _everything_ - the actual data, the sensitive features, the control features, and any sample parameters

The return values are also lighter on the Pythonic niceites, skipping the handling of the 'user supplied a single metric function' case.

Essentially, this turns `MetricFrame` into a wrapper for `DisaggregatedResult`, handling a wider variety of inputs from the user, and also 'cleaning up' some return values. In due course, I anticipate that this extraction will simplify the addition of bootstrapping to `MetricFrame`.

This is not intended to be user-facing (at least not at the current time), so documentation is light.

## Tests

- [ ] no new tests required
- [ ] new tests added
- [ ] existing tests adjusted

## Documentation

- [x] no documentation changes needed
- [ ] user guide added or updated
- [ ] API docs added or updated
- [ ] example notebook added or updated

### CODE REVIEW

**Comment by riedgar-ms** [comment]
My basic idea is that for bootstrapping, the `MetricFrame` will store an internal list of `DisaggregatedResult`s constructed from the bootstrapped evaluations, and then compute the required quantities on-demand.

**Comment by riedgar-ms** [comment]
@romanlutz @adrinjalali @MiroDudik ?

**Comment by riedgar-ms** [comment]
Ping @fairlearn/fairlearn-maintainers ?

**Review by adrinjalali — COMMENTED** [review]
Seems like this is not changing public API, so easy, but we should better document the new private API before merging.

**Comment by riedgar-ms** [comment]
> Seems like this is not changing public API, so easy, but we should better document the new private API before merging.

@adrinjalali I have improved the commenting :-)

**Comment by riedgar-ms** [comment]
@adrinjalali ? @MiroDudik ?

File: docs/user_guide/assessment.rst
Comment (adrinjalali) [inline docs/user_guide/assessment.rst:244]: so we ARE changing things, but I think I like these changes.

File: fairlearn/metrics/_disaggregated_metric.py
Comment (adrinjalali) [inline fairlearn/metrics/_disaggregated_metric.py:68]: would be nice to have the `Parameters` section here for the inputs and hat they are supposed to be.

File: fairlearn/metrics/_disaggregated_metric.py
Comment (adrinjalali) [inline fairlearn/metrics/_disaggregated_metric.py:95]: explain the input and output here as well. Same for the methods bellow.

**Review by adrinjalali — COMMENTED** [review]
Seems like we've got some validation we're not testing mostly.

I like the refactoring otherwise.

File: docs/user_guide/assessment.rst
Comment (riedgar-ms) [reply docs/user_guide/assessment.rst:244]: I'm not sure why that ended up changing. It was a slight surprise.

Comment (riedgar-ms) [reply docs/user_guide/assessment.rst:244]: Given Python's approach to typing, I was actually treating that as not-a-change :-)

File: fairlearn/metrics/_disaggregated_metric.py
Comment (riedgar-ms) [reply fairlearn/metrics/_disaggregated_metric.py:77]: Done

File: fairlearn/metrics/_disaggregated_metric.py
Comment (riedgar-ms) [reply fairlearn/metrics/_disaggregated_metric.py]: Done

File: docs/user_guide/assessment.rst
Comment (riedgar-ms) [reply docs/user_guide/assessment.rst:244]: And seeing other little doc issues pop up, this might even be due to a new Pandas release. I wish I knew :-/

**Comment by riedgar-ms** [comment]
OK.... docs updated and build is passing again...

**Comment by riedgar-ms** [comment]
@adrinjalali I have fixed the code coverage holes :-)

**Review by adrinjalali — APPROVED** [review]
Thanks @riedgar-ms

**Comment by adrinjalali** [comment]
I'm happy for this to be merged, not sure if we @MiroDudik wants to have a look or not.

**Pull request merged**
